src/backend/distributed/executor/executor.c

```diff
@@ -617,6 +623,24 @@ DistributedStatementRequiresRollback(DistributedPlan *distributedPlan)
 }
 
 
+/*
+ * LockPartitionsForDistributedPlan ensures commands take locks on all partitions
+ * of a distributed table.
+ */
+static void
+LockPartitionsForDistributedPlan(DistributedPlan *distributedPlan)
+{
+	if (distributedPlan->operation != CMD_SELECT)
+	{
+		Oid targetRelationId = distributedPlan->targetRelationId;
+
+		LockPartitionsInRelationList(list_make1_oid(targetRelationId), RowExclusiveLock);
+	}
+
+	LockPartitionsInRelationList(distributedPlan->relationIdList, AccessShareLock);
```

Comment:
Minor: Might be good to carry over the comment from the router executor for `AccessShareLock`s :
```
	/*
	 * We are taking locks on partitions of partitioned tables. These locks are
	 * necessary for locking tables that appear in the SELECT part of the query.
	 */
```